runtime/drivers/blob/blobdownloader.go

```diff
@@ -293,84 +262,60 @@ func (it *blobIterator) NextBatch(n int) ([]string, error) {
 			ext := filepath.Ext(obj.obj.Key)
 			partialReader, isPartialDownloadSupported := _partialDownloadReaders[ext]
 			downloadFull := obj.full || !isPartialDownloadSupported
 
 			// Collect metrics of download size and time
 			startTime := time.Now()
-			defer func() {
-				size := obj.obj.Size
-				st, err := file.Stat()
-				if err == nil {
-					size = st.Size()
+			if downloadFull {
+				// download full file
+				err = downloadObject(grpCtx, it.bucket, obj.obj.Key, file)
+			} else {
+				// download partial file
+				// check if, for smaller size we can download entire file
+				switch partialReader {
+				case "parquet":
+					err = downloadParquet(grpCtx, it.bucket, obj.obj, obj.extractOption, file)
+				case "csv":
+					err = downloadText(grpCtx, it.bucket, obj.obj, &textExtractOption{extractOption: obj.extractOption, hasCSVHeader: true}, file)
+				case "json":
+					err = downloadText(grpCtx, it.bucket, obj.obj, &textExtractOption{extractOption: obj.extractOption, hasCSVHeader: false}, file)
+				default:
+					// should not reach here
+					panic(fmt.Errorf("partial download not supported for extension %q", ext))
 				}
+			}
 
-				duration := time.Since(startTime)
-				it.logger.Info("download complete", zap.String("object", obj.obj.Key), zap.Duration("duration", duration), observability.ZapCtx(it.ctx))
-				drivers.RecordDownloadMetrics(grpCtx, &drivers.DownloadMetrics{
-					Connector: "blob",
-					Ext:       ext,
-					Partial:   !downloadFull,
-					Duration:  duration,
-					Size:      size,
-				})
-			}()
-
-			// download full file
-			if downloadFull {
-				return downloadObject(grpCtx, it.bucket, obj.obj.Key, file)
+			if err != nil {
+				// found an error stop triggering download of other files
+				stop = true
+				return err
 			}
-			// download partial file
-			// check if, for smaller size we can download entire file
-			switch partialReader {
-			case "parquet":
-				return downloadParquet(grpCtx, it.bucket, obj.obj, obj.extractOption, file)
-			case "csv":
-				return downloadText(grpCtx, it.bucket, obj.obj, &textExtractOption{extractOption: obj.extractOption, hasCSVHeader: true}, file)
-			case "json":
-				return downloadText(grpCtx, it.bucket, obj.obj, &textExtractOption{extractOption: obj.extractOption, hasCSVHeader: false}, file)
-			default:
-				// should not reach here
-				panic(fmt.Errorf("partial download not supported for extension %q", ext))
+
+			duration := time.Since(startTime)
+			size := obj.obj.Size
+			st, err := file.Stat()
+			if err == nil {
+				size = st.Size()
 			}
+			it.logger.Info("download complete", zap.String("object", obj.obj.Key), zap.Duration("duration", duration), observability.ZapCtx(it.ctx))
+			drivers.RecordDownloadMetrics(grpCtx, &drivers.DownloadMetrics{
+				Connector: "blob",
+				Ext:       ext,
+				Partial:   !downloadFull,
+				Duration:  duration,
+				Size:      size,
+			})
+			it.downloadedFiles <- downloadedFile{fileName: filename, sizeInBytes: size}
```

Comment:
If something goes wrong during insertion and the client calls `Close`, and the channel buffer is full, the background downloaders will hang forever while attempting to send to `downloadedFiles` here, because it is not consumed anymore.

Reply:
I don't think you can check a ctx while sending to a channel. You may need to drain the channel instead during close.